### ISSUE TITLE
'getPageProperties' mutating input parameters (polluting data external to SDK)

### ISSUE THREAD

**Issue by BMacLinden** [rudderlabs/rudder-sdk-js#956]
We discovered the Rudderstack sdk was polluting data that was passed in as an input parameter to event calls.  
In our case we were using this data object in other calls outside of Rudderstack, but your sdk had mutated it's original value.
_Steps to reproduce below._

https://github.com/rudderlabs/rudder-sdk-js/blob/6dfefe53f199c2dd7f70f60eb0719d99bf12e7f5/src/core/analytics.js#L880

**Steps to reproduce:**
- Create web page that includes Rudderstack JS SDK
- Create an object of properties you want to include with an event to Rudderstack: `const myEventData = { myDataOne: "some info", myDataTwo: "some other info" };`
- Fire Rudderstack event:`window.rudderanalytics.page("TEST", "Please do not pollute my data", myEventData);`
- Now use myEventData for some other purpose.  For test case, you can just log to console: `console.log("What happened to myEventData?", myEventData);`

**Results:**
- myEventData in console now looks like this:
`{
  "myDataOne": "some info",
  "myDataTwo": "some other info",
  "name": "name",
  "category": "category",
  "path": "path",
  "referrer": "referrer",
  "referring_domain": "referring_domain",
  "search": "",
  "title": "title",
  "url": "url",
  "tab_url": "tab_url,
  "initial_referrer": "initial_referrer",
  "initial_referring_domain": "initial_referring_domain"
}`

**Expected results:**
- Rudderstack sdk should not be modifying input parameters in a way that affects external data.
- myEventData should still look like: `{ myDataOne: "some info", myDataTwo: "some other info" }`

**Recommended fix:**
```
getPageProperties(properties, options) {
    const propsCopy = { ...properties };
    const defaultPageProperties = getDefaultPageProperties();
    const optionPageProperties = (options && options.page) || {};
    for (const key in defaultPageProperties) {
      if (propsCopy[key] === undefined) {
        propsCopy[key] = optionPageProperties[key] || defaultPageProperties[key];
      }
    }
    return propsCopy;
}
```

**Comment by MoumitaM** [comment]
HI @BMacLinden, Thank you for raising this. We will look into it and get back to you in some time.

**Comment by bardisg** [comment]
Hello @BMacLinden,

Thank you for raising this issue. We will apply fix as soon as possible and also ensure no other mutations take place in any user provided data/objects across the SDK.

As an interim solution for the next couple of days untill we release the fix, in case you are blocked, you can provide a deep clone or a shallow copy (if no nested objects) as input value.

Apologies for the incovenience.

**Comment by MoumitaM** [comment]
Hello @BMacLinden, We have released a fix in the latest version(2.28.0). Do try it and let us know feedback.

**Comment by BMacLinden** [comment]
Just did a quick check and looks good.  Thanks for the quick turnaround!
